packages/devkit/src/generators/project-name-directory-utils.ts

```diff
@@ -0,0 +1,282 @@
+import { prompt } from 'enquirer';
+import type { ProjectType } from 'nx/src/config/workspace-json-project-json';
+import type { Tree } from 'nx/src/generators/tree';
+import { requireNx } from '../../nx';
+import {
+  extractLayoutDirectory,
+  getWorkspaceLayout,
+} from '../utils/get-workspace-layout';
+import { names } from '../utils/names';
+
+const { joinPathFragments, readJson, readNxJson } = requireNx();
+
+export type ProjectNameDirectoryFormat = 'as-provided' | 'derived';
+export type ProjectGenerationOptions = {
+  name: string;
+  projectType: ProjectType;
+  directory?: string;
+  importPath?: string;
+  nameDirectoryFormat?: ProjectNameDirectoryFormat;
+  rootProject?: boolean;
+};
+
+export type ProjectNamesAndDirectories = {
```

Comment:
nitpick: This should match the function below.. so either `ProjectOptions` or `ProjectNameAndRoot`